File: web-app/src/app/service/tag.service.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { Message } from '../pojo/Message';
import { Page } from '../pojo/Page';
import { Tag } from '../pojo/Tag';

const tag_uri = '/tag';

@Injectable({
  providedIn: 'root'
})
export class TagService {
  constructor(private http: HttpClient) {}

  public loadTags(
    search: string | undefined,
    type: number | undefined,
    pageIndex: number,
    pageSize: number
  ): Observable<Message<Page<Tag>>> {
    pageIndex = pageIndex ? pageIndex : 0;
    pageSize = pageSize ? pageSize : 8;
    // HttpParams is unmodifiable, so we need to save the return value of append/set
    let httpParams = new HttpParams();
    httpParams = httpParams.appendAll({
      pageIndex: pageIndex,
      pageSize: pageSize
    });
    if (type != undefined) {
      httpParams = httpParams.append('type', type);
    }
    if (search != undefined && search != '' && search.trim() != '') {
      httpParams = httpParams.append('search', search.trim());
    }
    const options = { params: httpParams };
    return this.http.get<Message<Page<Tag>>>(tag_uri, options);
  }

  public newTags(body: Tag[]): Observable<Message<any>> {
    return this.http.post<Message<any>>(tag_uri, body);
  }

  public newTag(body: Tag): Observable<Message<any>> {
    const tags = [];
    tags.push(body);
    return this.http.post<Message<any>>(tag_uri, tags);
  }

  public editTag(body: Tag): Observable<Message<any>> {
    return this.http.put<Message<any>>(tag_uri, body);
  }

  public deleteTags(tagIds: Set<number>): Observable<Message<any>> {
    let httpParams = new HttpParams();
    tagIds.forEach(tagId => {
      // HttpParams is unmodifiable, so we need to save the return value of append/set
      // Method append can append same key, set will replace the previous value
      httpParams = httpParams.append('ids', tagId);
    });
    const options = { params: httpParams };
    return this.http.delete<Message<any>>(tag_uri, options);
  }
}

File: manager/src/main/java/org/apache/hertzbeat/manager/service/TagService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.hertzbeat.manager.service;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.apache.hertzbeat.common.entity.manager.Monitor;
import org.apache.hertzbeat.common.entity.manager.Tag;
import org.springframework.data.domain.Page;

/**
 * tag service
 */
public interface TagService {

    /**
     * new tags
     * @param tags tag
     */
    void addTags(List<Tag> tags);

    /**
     * update tag
     * @param tag Tag
     */
    void modifyTag(Tag tag);

    /**
     * get tag page list
     * @param search        Tag content search
     * @param type          Tag type
     * @param pageIndex     List current page
     * @param pageSize      Number of list pagination
     * @return Tags
     */
    Page<Tag> getTags(String search, Byte type, int pageIndex, int pageSize);

    /**
     * delete tags
     * @param ids tag id list
     */
    void deleteTags(HashSet<Long> ids);

    /**
     * list tags
     * @param ids tag id list
     * @return tag list
     */
    List<Tag> listTag(Set<Long> ids);

    /**
     * remove monitor system tags
     * @param monitor monitor
     */
    void deleteMonitorSystemTags(Monitor monitor);
}
